File: auto_fetcher.py
import subprocess, json, time, base64, os, signal, asyncio
import requests, websockets

GAME_URL   = "https://68gbvn88.bar/"
BOT_SERVER = os.environ.get("BOT_SERVER", "http://localhost:8080/")

async def fetch_token() -> bool:
    print("🚀 [FETCH] Khởi động Chrome headless...")

    # Kill chrome cũ (nếu còn)
    subprocess.run(["pkill", "-9", "-f", "google-chrome"], stderr=subprocess.DEVNULL)

    proc = subprocess.Popen([
        "google-chrome",
        "--headless=new",
        "--remote-debugging-port=9222",
        "--no-sandbox",
        "--disable-dev-shm-usage",
        "--disable-gpu",
        "--disable-extensions",
        "--user-agent=Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/137.0.0.0 Safari/537.36",
        "--user-data-dir=/tmp/bot_chrome_profile",
        GAME_URL
    ], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)

    # Đợi Chrome mở cổng debug (tối đa 15s)
    ws_url = None
    for i in range(15):
        await asyncio.sleep(1)
        try:
            pages = requests.get("http://localhost:9222/json", timeout=2).json()
            ws_url = next((p["webSocketDebuggerUrl"] for p in pages if p["type"] == "page"), None)
            if ws_url:
                break
        except Exception:
            if i % 3 == 0:
                print(f"⏳ [FETCH] Đợi Chrome... ({i}s)")

    if not ws_url:
        print("❌ [FETCH] Chrome không phản hồi sau 15s.")
        _kill(proc)
        return False

    try:
        async with websockets.connect(ws_url) as ws:
            for cmd_id, method in [(1, "Network.enable"), (2, "Page.enable"), (3, "Runtime.enable")]:
                await ws.send(json.dumps({"id": cmd_id, "method": method}))

            print("📡 [FETCH] Đang đợi game tải và bắt token...")
            start = time.time()

            while time.time() - start < 180:
                try:
                    data = json.loads(await asyncio.wait_for(ws.recv(), timeout=2.0))
                except asyncio.TimeoutError:
                    # Thử click để phá popup
                    elapsed = int(time.time() - start)
                    if 15 < elapsed < 60 and elapsed % 10 == 0:
                        for x, y in [(600, 500), (400, 700), (800, 700)]:
                            for ev in ("mousePressed", "mouseReleased"):
                                await ws.send(json.dumps({
                                    "id": 9000 + x,
                                    "method": "Input.dispatchMouseEvent",
                                    "params": {"type": ev, "x": x, "y": y, "button": "left", "clickCount": 1}
                                }))
                        print(f"🖱️ [FETCH] Click phá popup ({elapsed}s)...")
                    continue

                # Bắt WebSocket frame gửi đi từ trang
                if data.get("method") == "Network.webSocketFrameSent":
                    payload = data["params"]["response"]["payloadData"]
                    try:
                        raw = base64.b64decode(payload)
                        if len(raw) > 50 and raw[0] == 0x04:
                            hex_str = "b'" + "".join(f"\\x{b:02x}" for b in raw) + "'"
                            requests.post(BOT_SERVER, json={"token": hex_str}, timeout=5)
                            print("✅ [FETCH] Token đã bắt và gửi về Bot!")
                            return True
                    except Exception:
                        pass

    except Exception as e:
        print(f"❌ [FETCH] Lỗi CDP: {e}")
    finally:
        _kill(proc)

    return False

def _kill(proc):
    try:
        os.kill(proc.pid, signal.SIGTERM)
        time.sleep(1)
    except Exception:
        pass
    subprocess.run(["pkill", "-9", "-f", "google-chrome"], stderr=subprocess.DEVNULL)

if __name__ == "__main__":
    asyncio.run(fetch_token())

File: auto_fetcher.js
const puppeteer = require('puppeteer'); // Dùng puppeteer (không phải core) để tự tải chromium
const fs = require('fs');

const GAME_URL = "https://68gbvn88.bar/";
const BOT_SERVER = process.env.BOT_SERVER || "http://localhost:8080/api/token";

async function fetchToken() {
    console.log("🚀 [FETCH-JS] Khởi động trình duyệt...");

    let browser;
    try {
        const isRender = process.env.RENDER || false;

        const launchOptions = {
            headless: "new",
            executablePath: process.env.PUPPETEER_EXECUTABLE_PATH || null,
            args: [
                '--no-sandbox',
                '--disable-setuid-sandbox',
                '--disable-dev-shm-usage',
                '--disable-gpu',
                '--single-process' // Tiết kiệm RAM trên Render
            ]
        };

        // Nếu ở local Linux của bác thì mới ép path Chrome
        if (fs.existsSync('/usr/bin/google-chrome')) {
            launchOptions.executablePath = '/usr/bin/google-chrome';
        }

        browser = await puppeteer.launch(launchOptions);
        const page = await browser.newPage();
        await page.setUserAgent("Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/137.0.0.0 Safari/537.36");

        // Bắt sự kiện WebSocket qua CDP
        const client = await page.target().createCDPSession();
        await client.send('Network.enable');

        let captured_ws_url = null;
        client.on('Network.webSocketCreated', (params) => {
            console.log(`🔗 [FETCH-JS] Phát hiện WebSocket: ${params.url}`);
            captured_ws_url = params.url;
        });

        client.on('Network.webSocketFrameSent', async (params) => {
            const payload = params.response.payloadData;
            const buffer = Buffer.from(payload, 'base64');

            if (buffer.length > 50 && buffer[0] === 0x04) {
                const hex = buffer.toString('hex');
                console.log("✅ [FETCH-JS] Bắt được Token! Đang gửi về Bot...");

                try {
                    const cookies = await page.cookies();
                    const cookieStr = cookies.map(c => `${c.name}=${c.value}`).join('; ');

                    const response = await fetch(BOT_SERVER, {
                        method: 'POST',
                        body: JSON.stringify({
                            token: hex,
                            ws_url: captured_ws_url,
                            cookies: cookieStr
                        }),
                        headers: { 'Content-Type': 'application/json' }
                    });
                    const res = await response.json();
                    console.log("🤖 [BOT] Phản hồi:", res);
                    if (res.status === "ok") {
                        console.log("🏁 [FINISH] Hoàn tất!");
                        process.exit(0);
                    }
                } catch (e) {
                    console.error("❌ [ERROR] Lỗi gửi token:", e.message);
                }
            }
        });

        console.log("📡 [FETCH-JS] Đang tải game...");
        await page.goto(GAME_URL, { waitUntil: 'networkidle2', timeout: 90000 });

        // Safety EXIT after 2 minutes if still stuck
        const safetyFinal = setTimeout(() => {
            console.log("💀 [CRITICAL] Token fetcher stuck too long. Exiting.");
            process.exit(1);
        }, 120000);

        // Click phá popup
        await page.mouse.click(600, 500);
        console.log("🖱️ [FETCH-JS] Đã click giả lập...");

        // Timeout 4 phút
        setTimeout(() => {
            console.log("⏰ [TIMEOUT] Không bắt được token sau 4 phút.");
            process.exit(1);
        }, 240000);

    } catch (err) {
        console.error("❌ [FETCH-JS] Lỗi:", err.message);
        process.exit(1);
    }
}

fetchToken();
